spec/controllers/tasks_controller_spec.rb

```diff
@@ -1043,17 +1043,16 @@
   end
 
   describe 'POST #generate_test' do
-    let(:task_user) { create(:user) }
+    let(:task_user) { create(:user, personal_openai_api_key: 'valid_api_key') }
     let(:access_level) { :public }
     let(:task) { create(:task, user: task_user, access_level:) }
+    let(:mock_client) { instance_double(OpenAI::Client) }
+    let(:mock_models) { instance_double(OpenAI::Models, list: {'data' => [{'id' => 'text-davinci-002'}]}) }
```

Comment:
You are testing an implementation detail of your service here. That should be tested in the spec for the service itself.
Instead you should be mocking the service here and create different context, where it behaves differently (raise error, etc)

Reply:
I am not sure I understand. Here I am testing that the validation of the key is correct. doesn't that belong here?

Reply:
Aha I see, I looked at the full code and understand what you mean. will remove them as they are not used here as you mentioned